tests/test_integration.py

```diff
@@ -3210,3 +3214,93 @@ def test_console_script_from_pex_path(tmpdir):
     result.assert_success()
 
     assert "hello world!\n" == subprocess.check_output(args=[pex_file]).decode("utf-8")
+
+
+def test_execute_module_issues_1018(tmpdir):
+    # type: (Any) -> None
+    src_dir = os.path.join(str(tmpdir), "src")
+    with safe_open(os.path.join(src_dir, "issues_1018.py"), "w") as fp:
+        fp.write(
+            dedent(
+                """\
+                import pickle
+
+                def add(a, b):
+                    return a + b
+
+                def main():
+                    pickle_add = pickle.dumps(add)
+                    add_clone = pickle.loads(pickle_add)
+                    print(add_clone(1, 2))
+
+                if __name__ == "__main__":
+                    main()
+                """
+            )
+        )
+    expected_output = b"3\n"
+
+    # There are 9 ways we can invoke the module above using Pex corresponding to the 2D matrix with
+    # axes: {zipapp, unzip, venv} x {entrypoint-function, entrypoint-module, ad-hoc (-m) module}.
+    # Of these, zipapp + entrypoint-module is the only combination where we can't both satisfy being
+    # able to re-exec a PEX based on argv[0] and support pickling.
```

Comment:
Could this test be phrased as a nested loop over these dimensions, and a conditional for the exceptional case? Or the case itself be extracted into a helper method instead?

Reply:
It probably could, but there really isn't a whole lot repeated or shared so it would get harder to read. The --venv --seed mode is an outlier from all other cases for one.